packages/service-library/tests/test_rabbitmq_robust_rpc.py

```diff
@@ -0,0 +1,229 @@
+
+    clients: list[RobustRPCClient] = []
+    for _ in range(MULTIPLE_REQUESTS_COUNT):
+        client = RobustRPCClient(rabbit_service)
+        clients.append(client)
+
+    # worst case scenario
+    await asyncio.gather(*[c.start() for c in clients])
+
+    requests: list[Awaitable] = []
+    expected_result: list[int] = []
+    for i in range(MULTIPLE_REQUESTS_COUNT):
+        client = clients[i]
+        requests.append(client.request(add_me.__name__, x=1 + i, y=2 + i))
+        expected_result.append(3 + i * 2)
+
+    assert await asyncio.gather(*requests) == expected_result
+
+    # worst case scenario
+    await asyncio.gather(*[c.stop() for c in clients])
+
+
+async def test_raise_error_if_not_started(rabbit_service: RabbitSettings):
+    client = RobustRPCClient(rabbit_settings=rabbit_service)
+    with pytest.raises(NotStartedError):
+        await client.request(add_me, x=1, y=2)
+
+    # expect not to raise error
+    await client.stop()
+
+    server = RobustRPCServer(rabbit_settings=rabbit_service)
+    with pytest.raises(NotStartedError):
+        await server.register(add_me)
+
+    # expect not to raise error
+    await server.stop()
+
+
+async def _assert_event_not_registered(rpc_client: RobustRPCClient):
+    with pytest.raises(RemoteMethodNotRegisteredError) as exec_info:
+        assert await rpc_client.request(add_me.__name__, x=1, y=3) == 3
+    assert (
+        f"Could not find a remote method named: '{add_me.__name__}'"
+        in f"{exec_info.value}"
+    )
+
+
+async def test_server_not_started(rpc_client: RobustRPCClient):
+    await _assert_event_not_registered(rpc_client)
+
+
+async def test_server_handler_not_registered(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await _assert_event_not_registered(rpc_client)
+
+
+async def test_request_is_missing_arguments(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    await rpc_server.register(add_me)
+
+    # missing 1 argument
+    with pytest.raises(TypeError) as exec_info:
+        await rpc_client.request(add_me.__name__, x=1)
+    assert (
+        f"{add_me.__name__}() missing 1 required keyword-only argument: 'y'"
+        in f"{exec_info.value}"
+    )
+
+    # missing all arguments
+    with pytest.raises(TypeError) as exec_info:
+        await rpc_client.request(add_me.__name__)
+    assert (
+        f"{add_me.__name__}() missing 2 required keyword-only arguments: 'x' and 'y'"
+        in f"{exec_info.value}"
+    )
+
+
+async def test_client_cancels_long_running_request_or_client_takes_too_much_to_respond(
+    rpc_client: RobustRPCClient, rpc_server: RobustRPCServer
+):
+    async def _long_running(*, time_to_sleep: float) -> None:
+        await asyncio.sleep(time_to_sleep)
+
+    await rpc_server.register(_long_running)
+
+    # this task will be cancelled
+    with pytest.raises(asyncio.TimeoutError):
+        await rpc_client.request(_long_running.__name__, time_to_sleep=10, timeout=0.5)
```

Comment:
Here I see timing out, not cancellation.
How does cancellation works out?
can you create a test with a asyncio task, then cancel it? I would expect the cancellation to propagate all the way.

Reply:
It is equivalent since the code behind the timeout is implemented with a cancel and the error is propagated to the client all the way.